### PR TITLE
[Feature/EP-23-2 blueprint] CSPF links metadata filtering

### PR DESCRIPTION
**Pull request by viniarck**
#### Summary

- This PR implemented the [EP-23-2 blueprint](https://github.com/kytos-ng/kytos/blob/master/docs/blueprints/EP023-2.rst) CSPF filtering capabilities. On this PR I continued the work that Marvin, David and Arturo contributed [on this previous PR](https://github.com/kytos/pathfinder/pull/68).
- Check out the [`CHANGELOG.rst` below or on this link](https://github.com/viniarck/pathfinder/blob/feature/ep_23_2/CHANGELOG.rst#230---2021-10-19) that summarizes the changes and additions on this PR. 

I've also explored this version on `mininet` with `mef_eline` and worked as intended since no breaking changes were expected, considering how [`mef_eline`](https://github.com/kytos-ng/mef_eline/blob/master/models.py#L101-L113) is currently interfacing with `pathfinder`, and also since it's using at most 2 paths, which was the `spf_max_paths` default value for now to avoid leaving too much room for high CPU usage when the graph is larger in terms of edges and nodes (which will probably be the case in production). @jab1982 @dgarc330 I think it would be neat to augment some end to end tests to indirectly test that with mef eline, and plan for it, we've got several unit and integration tests, but having the end-to-end would also be great and less biased since even though most of the tests were contributed by the previous authors, I unified some of them. 

#### Perf

When I was exploring some larger topologies in the integration tests, I noticed in some cases, it was noticeable hanging so I decided to benchmark and the function we used to use, as it was being invoked, was exhaustively computing all the best paths,  `test_n_shortest_simple_paths` below,  and now we have the k paths being computed lazily, represented by the `test_k*_shortest_simple_paths`  test functions, and these ones are optimal based on the k input:

```
❯ pytest bench.py --benchmark-columns='min,max,mean,stddev,rounds,iterations'                                                                                                   [NORMAL]
================================================================================== test session starts ===================================================================================
platform linux -- Python 3.6
benchmark: 3.4.1 (defaults: timer=time.perf_counter disable_gc=False min_rounds=5 min_time=0.000005 max_time=1.0 calibration_precision=10 warmup=False warmup_iterations=100000)
rootdir: /mnt/b/repos/networkx_proto
plugins: benchmark-3.4.1
collected 6 items

bench.py ......                                                                                                                                                                    [100%]


-------------------------------------------------------------- benchmark: 6 tests -------------------------------------------------------------
Name (time in us)                         Min                    Max                   Mean                StdDev            Rounds  Iterations
-----------------------------------------------------------------------------------------------------------------------------------------------
test_k1_shortest_simple_paths         40.4670 (1.0)         473.0600 (3.45)         44.1140 (1.0)          9.6005 (1.12)      14698           1
test_dijkstra                         45.7080 (1.13)        137.0790 (1.0)          49.2101 (1.12)         8.5580 (1.0)       12730           1
test_k2_shortest_simple_paths         52.4290 (1.30)        466.3990 (3.40)         55.3048 (1.25)         8.7324 (1.02)      11143           1
test_k4_shortest_simple_paths         92.0600 (2.27)        295.0550 (2.15)         99.4376 (2.25)        16.7802 (1.96)       8046           1
test_n_shortest_simple_paths      55,771.9020 (>1000.0)  59,501.4840 (434.07)   56,846.8905 (>1000.0)    923.8802 (107.96)       18           1
test_all_shortest_paths           56,784.1620 (>1000.0)  60,191.6640 (439.10)   57,836.4018 (>1000.0)  1,090.3481 (127.41)       18           1
-----------------------------------------------------------------------------------------------------------------------------------------------

Legend:
  Outliers: 1 Standard Deviation from Mean; 1.5 IQR (InterQuartile Range) from 1st Quartile and 3rd Quartile.
  OPS: Operations Per Second, computed as 1 / Mean
=================================================================================== 6 passed in 6.79s ====================================================================================
```

#### UI 

2021-10-19-191242_1919x950_scrot

I used the exact same components and layout, I'm not a front-end specialist, but one of the things we could improve in terms of UX would be to more easily to click on desired/undesired links, it's usable as it is, but maybe something to align and prioritize later on.

#### Open API spec preview

@ajoaoff here's the openapi spec preview, it's still backwards compatible from `mef_eline`'s perspective, if you need help integrating this let me know. thanks. 

2021-10-19-181933_889x899_scrot

### CODE REVIEW

File: openapi.yml [hunk 1]
```diff
@@ -4,89 +4,220 @@ info:
...
+                    - $ref: "#/components/schemas/Attributes"
+                  example:
+                    delay: 81
+                    utilization: 100
+                    reliability: 3
+                minimum_flexible_hits:
+                  type: integer
+                  description: Minimum number of attributes listed in flexible_metrics that a path will meet.
+                  example: 2
+                  minimum: 1
```
[the remark this answers is not in the document]
Comment (viniarck) [reply openapi.yml]: for sure, done, applied the suggested commit.

File: openapi.yml [hunk 2]
```diff
@@ -4,89 +4,220 @@ info:
...
+          description: Path cost
+          example: 100
         hops:
-          type: array
-          description: Hops between two points
-          items:
-            $ref: "#/components/schemas/Hop"
-          example:
-            $ref: "#/examples/Hops"
+          allOf:
```
[the remark this answers is not in the document]
Comment (viniarck) [reply openapi.yml]: I don't think so, ended up grabbing this part from the previous PR, indeed it looks redundant. I removed it and tried on [swagger editor](https://editor.swagger.io) and it renders correctly. Done here https://github.com/kytos-ng/pathfinder/pull/1/commits/1cb5e67593363483895292f2affd7bea0876dbdb

File: tests/integration/test_paths_edges.py [hunk 3]
```diff
@@ -0,0 +1,249 @@
...
+        self.initializer()
+
+        for source, destination in combos:
+            paths = self.graph.constrained_k_shortest_paths(
+                source, destination
+            )
+            assert paths
+            for path in paths:
+                assert path["hops"][0] == source
+                assert path["hops"][-1] == destination
```
[the remark this answers is not in the document]
Comment (viniarck) [reply tests/integration/test_paths_edges.py]: done https://github.com/kytos-ng/pathfinder/commit/4e7e19b071d164ead7ea8b952cdee381aa145ecc

File: openapi.yml [hunk 4]
```diff
@@ -4,89 +4,220 @@ info:
...
-                    type: array
-                    items:
-                      $ref: "#/components/schemas/Path"
+                    $ref: "#/components/schemas/Paths"
+        400:
+          description: Illegal value provided.
+          content:
+            application/json:
+              schema:
+                $ref: "#/components/schemas/BadRequest"
```
[the remark this answers is not in the document]
Comment (ajoaoff) [reply openapi.yml:139]: 204 would make sense only if nothing is returned. What I meant was if it should return a 404 when there is no path, but you are right, 404 is used when you want an specific object. In this specific case, returning an empty list is ok.

**Review by ajoaoff — APPROVED** [review]
LGTM

**Comment by viniarck** [comment]
> LGTM

I appreciated your review @ajoaoff, thanks, I'll merge this PR today.

**Pull request merged**